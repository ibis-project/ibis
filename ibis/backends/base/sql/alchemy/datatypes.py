from __future__ import annotations

from typing import TYPE_CHECKING

import sqlalchemy as sa
import sqlalchemy.types as sat
import toolz
from sqlalchemy.ext.compiler import compiles

import ibis.expr.datatypes as dt
from ibis.backends.base.sql.alchemy.geospatial import geospatial_supported
from ibis.backends.base.sqlglot.datatypes import SqlglotType
from ibis.common.collections import FrozenDict
from ibis.formats import TypeMapper

if TYPE_CHECKING:
    from collections.abc import Mapping

if geospatial_supported:
    import geoalchemy2 as ga


class ArrayType(sat.UserDefinedType):
    def __init__(self, value_type: sat.TypeEngine):
        self.value_type = sat.to_instance(value_type)

    def result_processor(self, dialect, coltype) -> None:
        if not coltype.lower().startswith("array"):
            return None

        inner_processor = (
            self.value_type.result_processor(dialect, coltype[len("array(") : -1])
            or toolz.identity
        )

        return lambda v: v if v is None else list(map(inner_processor, v))


@compiles(ArrayType, "default")
def compiles_array(element, compiler, **kw):
    return f"ARRAY({compiler.process(element.value_type, **kw)})"


class StructType(sat.UserDefinedType):
    cache_ok = True

    def __init__(self, fields: Mapping[str, sat.TypeEngine]) -> None:
        self.fields = FrozenDict(
            {name: sat.to_instance(typ) for name, typ in fields.items()}
        )


@compiles(StructType, "default")
def compiles_struct(element, compiler, **kw):
    quote = compiler.dialect.identifier_preparer.quote
    content = ", ".join(
        f"{quote(field)} {compiler.process(typ, **kw)}"
        for field, typ in element.fields.items()
    )
    return f"STRUCT({content})"


class MapType(sat.UserDefinedType):
    def __init__(self, key_type: sat.TypeEngine, value_type: sat.TypeEngine):
        self.key_type = sat.to_instance(key_type)
        self.value_type = sat.to_instance(value_type)


@compiles(MapType, "default")
def compiles_map(element, compiler, **kw):
    key_type = compiler.process(element.key_type, **kw)
    value_type = compiler.process(element.value_type, **kw)
    return f"MAP({key_type}, {value_type})"


class UInt64(sat.Integer):
    pass


class UInt32(sat.Integer):
    pass


class UInt16(sat.Integer):
    pass


class UInt8(sat.Integer):
    pass


@compiles(UInt64, "postgresql")
@compiles(UInt32, "postgresql")
@compiles(UInt16, "postgresql")
@compiles(UInt8, "postgresql")
@compiles(UInt64, "mssql")
@compiles(UInt32, "mssql")
@compiles(UInt16, "mssql")
@compiles(UInt8, "mssql")
@compiles(UInt64, "mysql")
@compiles(UInt32, "mysql")
@compiles(UInt16, "mysql")
@compiles(UInt8, "mysql")
@compiles(UInt64, "snowflake")
@compiles(UInt32, "snowflake")
@compiles(UInt16, "snowflake")
@compiles(UInt8, "snowflake")
@compiles(UInt64, "sqlite")
@compiles(UInt32, "sqlite")
@compiles(UInt16, "sqlite")
@compiles(UInt8, "sqlite")
@compiles(UInt64, "trino")
@compiles(UInt32, "trino")
@compiles(UInt16, "trino")
@compiles(UInt8, "trino")
def compile_uint(element, compiler, **kw):
    dialect_name = compiler.dialect.name
    raise TypeError(
        f"unsigned integers are not supported in the {dialect_name} backend"
    )


try:
    UUID = sat.UUID
except AttributeError:

    class UUID(sat.String):
        pass

else:

    @compiles(UUID, "default")
    def compiles_uuid(element, compiler, **kw):
        return "UUID"


class Unknown(sa.Text):
    pass


_from_sqlalchemy_types = {
    sat.BOOLEAN: dt.Boolean,
    sat.Boolean: dt.Boolean,
    sat.BINARY: dt.Binary,
    sat.LargeBinary: dt.Binary,
    sat.DATE: dt.Date,
    sat.Date: dt.Date,
    sat.TEXT: dt.String,
    sat.Text: dt.String,
    sat.TIME: dt.Time,
    sat.Time: dt.Time,
    sat.VARCHAR: dt.String,
    sat.CHAR: dt.String,
    sat.String: dt.String,
    sat.SMALLINT: dt.Int16,
    sat.SmallInteger: dt.Int16,
    sat.INTEGER: dt.Int32,
    sat.Integer: dt.Int32,
    sat.BIGINT: dt.Int64,
    sat.BigInteger: dt.Int64,
    sat.REAL: dt.Float32,
    sat.FLOAT: dt.Float64,
    UInt16: dt.UInt16,
    UInt32: dt.UInt32,
    UInt64: dt.UInt64,
    UInt8: dt.UInt8,
    Unknown: dt.Unknown,
    sat.JSON: dt.JSON,
    UUID: dt.UUID,
}

_to_sqlalchemy_types = {
    dt.Null: sat.NullType,
    dt.Date: sat.Date,
    dt.Time: sat.Time,
    dt.Boolean: sat.Boolean,
    dt.Binary: sat.LargeBinary,
    dt.String: sat.Text,
    dt.Decimal: sat.Numeric,
    # Mantissa-based
    dt.Float16: sat.REAL,
    dt.Float32: sat.REAL,
    # precision is the number of bits in the mantissa
    # without specifying this, some backends interpret the type as FLOAT, which
    # means float32 (and precision == 24)
    dt.Float64: sat.FLOAT(precision=53),
    dt.Int8: sat.SmallInteger,
    dt.Int16: sat.SmallInteger,
    dt.Int32: sat.Integer,
    dt.Int64: sat.BigInteger,
    dt.UInt8: UInt8,
    dt.UInt16: UInt16,
    dt.UInt32: UInt32,
    dt.UInt64: UInt64,
    dt.JSON: sat.JSON,
    dt.Interval: sat.Interval,
    dt.Unknown: Unknown,
    dt.MACADDR: sat.Text,
    dt.INET: sat.Text,
    dt.UUID: UUID,
}

_FLOAT_PREC_TO_TYPE = {
    11: dt.Float16,
    24: dt.Float32,
    53: dt.Float64,
}

_GEOSPATIAL_TYPES = {
    "POINT": dt.Point,
    "LINESTRING": dt.LineString,
    "POLYGON": dt.Polygon,
    "MULTILINESTRING": dt.MultiLineString,
    "MULTIPOINT": dt.MultiPoint,
    "MULTIPOLYGON": dt.MultiPolygon,
    "GEOMETRY": dt.Geometry,
    "GEOGRAPHY": dt.Geography,
}


class AlchemyType(TypeMapper):
    @classmethod
    def to_string(cls, dtype: dt.DataType):
        dialect_class = sa.dialects.registry.load(cls.dialect)
        return str(
            sa.types.to_instance(cls.from_ibis(dtype)).compile(dialect=dialect_class())
        )

    @classmethod
    def from_string(cls, type_string, nullable=True):
        return SqlglotType.from_string(type_string, nullable=nullable)

    @classmethod
    def from_ibis(cls, dtype: dt.DataType) -> sat.TypeEngine:
        """Convert an Ibis type to a SQLAlchemy type.

        Parameters
        ----------
        dtype
            Ibis type to convert.

        Returns
        -------
        SQLAlchemy type.
        """
        if dtype.is_decimal():
            return sat.NUMERIC(dtype.precision, dtype.scale)
        elif dtype.is_timestamp():
            return sat.TIMESTAMP(timezone=bool(dtype.timezone))
        elif dtype.is_array():
            return ArrayType(cls.from_ibis(dtype.value_type))
        elif dtype.is_struct():
            fields = {k: cls.from_ibis(v) for k, v in dtype.fields.items()}
            return StructType(fields)
        elif dtype.is_map():
            return MapType(
                cls.from_ibis(dtype.key_type), cls.from_ibis(dtype.value_type)
            )
        elif dtype.is_geospatial():
            if geospatial_supported:
                if dtype.geotype == "geometry":
                    return ga.Geometry
                elif dtype.geotype == "geography":
                    return ga.Geography
                else:
                    return ga.types._GISType
            else:
                raise TypeError("geospatial types are not supported")
        else:
            return _to_sqlalchemy_types[type(dtype)]

    @classmethod
    def to_ibis(cls, typ: sat.TypeEngine, nullable: bool = True) -> dt.DataType:
        """Convert a SQLAlchemy type to an Ibis type.

        Parameters
        ----------
        typ
            SQLAlchemy type to convert.
        nullable : bool, optional
            Whether the returned type should be nullable.

        Returns
        -------
        Ibis type.
        """
        if dtype := _from_sqlalchemy_types.get(type(typ)):
            return dtype(nullable=nullable)
        elif isinstance(typ, sat.Float):
            if (float_typ := _FLOAT_PREC_TO_TYPE.get(typ.precision)) is not None:
                return float_typ(nullable=nullable)
            return dt.Decimal(typ.precision, typ.scale, nullable=nullable)
        elif isinstance(typ, sat.Numeric):
            return dt.Decimal(typ.precision, typ.scale, nullable=nullable)
        elif isinstance(typ, ArrayType):
            return dt.Array(cls.to_ibis(typ.value_type), nullable=nullable)
        elif isinstance(typ, sat.ARRAY):
            ndim = typ.dimensions
            if ndim is not None and ndim != 1:
                raise NotImplementedError("Nested array types not yet supported")
            return dt.Array(cls.to_ibis(typ.item_type), nullable=nullable)
        elif isinstance(typ, StructType):
            fields = {k: cls.to_ibis(v) for k, v in typ.fields.items()}
            return dt.Struct(fields, nullable=nullable)
        elif isinstance(typ, MapType):
            return dt.Map(
                cls.to_ibis(typ.key_type),
                cls.to_ibis(typ.value_type),
                nullable=nullable,
            )
        elif isinstance(typ, sa.DateTime):
            timezone = "UTC" if typ.timezone else None
            return dt.Timestamp(timezone, nullable=nullable)
        elif isinstance(typ, sat.String):
            return dt.String(nullable=nullable)
        elif geospatial_supported and isinstance(typ, ga.types._GISType):
            name = typ.geometry_type.upper()
            try:
                return _GEOSPATIAL_TYPES[name](geotype=typ.name, nullable=nullable)
            except KeyError:
                raise ValueError(f"Unrecognized geometry type: {name}")
        else:
            raise TypeError(f"Unable to convert type: {typ!r}")
