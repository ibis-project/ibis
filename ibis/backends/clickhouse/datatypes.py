from __future__ import annotations

from typing import Literal

import sqlglot.expressions as sge

import ibis
import ibis.expr.datatypes as dt
from ibis.backends.base.sqlglot.datatypes import SqlglotType
from ibis.common.collections import FrozenDict

typecode = sge.DataType.Type


# TODO(kszucs): add a bool converter method to support different clickhouse bool types
def _bool_type() -> Literal["Bool", "UInt8", "Int8"]:
    return getattr(getattr(ibis.options, "clickhouse", None), "bool_type", "Bool")


class ClickhouseType(SqlglotType):
    dialect = "clickhouse"
    default_decimal_precision = None
    default_decimal_scale = None
    default_nullable = False

    unknown_type_strings = FrozenDict(
        {
            "ipv4": dt.INET(nullable=default_nullable),
            "ipv6": dt.INET(nullable=default_nullable),
            "object('json')": dt.JSON(nullable=default_nullable),
            "array(null)": dt.Array(dt.null, nullable=default_nullable),
            "array(nothing)": dt.Array(dt.null, nullable=default_nullable),
        }
    )

    @classmethod
    def from_ibis(cls, dtype: dt.DataType) -> sge.DataType:
        """Convert a sqlglot type to an ibis type."""
        typ = super().from_ibis(dtype)
        if dtype.nullable and not dtype.is_map():
            # map cannot be nullable in clickhouse
            return sge.DataType(this=typecode.NULLABLE, expressions=[typ])
        else:
            return typ

    @classmethod
    def _from_sqlglot_NULLABLE(cls, inner_type: sge.DataType) -> dt.DataType:
        return cls.to_ibis(inner_type, nullable=True)

    @classmethod
    def _from_sqlglot_DATETIME(
        cls, timezone: sge.DataTypeParam | None = None
    ) -> dt.Timestamp:
        return dt.Timestamp(
            scale=0,
            timezone=None if timezone is None else timezone.this.this,
            nullable=cls.default_nullable,
        )

    @classmethod
    def _from_sqlglot_DATETIME64(
        cls,
        scale: sge.DataTypeSize | None = None,
        timezone: sge.Literal | None = None,
    ) -> dt.Timestamp:
        return dt.Timestamp(
            timezone=None if timezone is None else timezone.this.this,
            scale=int(scale.this.this),
            nullable=cls.default_nullable,
        )

    @classmethod
    def _from_sqlglot_LOWCARDINALITY(cls, inner_type: sge.DataType) -> dt.DataType:
        return cls.to_ibis(inner_type)

    @classmethod
    def _from_sqlglot_NESTED(cls, *fields: sge.DataType) -> dt.Struct:
        fields = {
            field.name: dt.Array(
                cls.to_ibis(field.args["kind"]), nullable=cls.default_nullable
            )
            for field in fields
        }
        return dt.Struct(fields, nullable=cls.default_nullable)

    @classmethod
    def _from_ibis_Timestamp(cls, dtype: dt.Timestamp) -> sge.DataType:
        if dtype.timezone is None:
            timezone = None
        else:
            timezone = sge.DataTypeParam(this=sge.Literal.string(dtype.timezone))

        if dtype.scale is None:
            return sge.DataType(this=typecode.DATETIME, expressions=[timezone])
        else:
            scale = sge.DataTypeParam(this=sge.Literal.number(dtype.scale))
            return sge.DataType(this=typecode.DATETIME64, expressions=[scale, timezone])

    @classmethod
    def _from_ibis_Map(cls, dtype: dt.Map) -> sge.DataType:
        # key cannot be nullable in clickhouse
        key_type = cls.from_ibis(dtype.key_type.copy(nullable=False))
        value_type = cls.from_ibis(dtype.value_type)
        return sge.DataType(this=typecode.MAP, expressions=[key_type, value_type])
