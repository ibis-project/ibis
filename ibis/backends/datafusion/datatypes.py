from __future__ import annotations

from ibis.backends.pyarrow.datatypes import to_pyarrow_type

__all__ = ("to_pyarrow_type",)
