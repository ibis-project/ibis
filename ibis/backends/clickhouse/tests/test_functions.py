from __future__ import annotations

import math
from datetime import date, datetime
from operator import methodcaller

import pandas as pd
import pandas.testing as tm
import pytest
from pytest import param

import ibis
import ibis.expr.datatypes as dt
import ibis.expr.types as ir
from ibis import literal as L
from ibis import udf

pytest.importorskip("clickhouse_connect")


@pytest.mark.parametrize("to_type", ["int8", "int16", "float32", "float", "!float64"])
def test_cast_double_col(alltypes, to_type, snapshot):
    expr = alltypes.double_col.cast(to_type)
    result = expr.compile()
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "to_type",
    [
        "int8",
        "int16",
        "!string",
        "timestamp",
        "date",
        "!map<string, int64>",
        "!struct<a: string, b: int64>",
    ],
)
def test_cast_string_col(alltypes, to_type, snapshot):
    expr = alltypes.string_col.cast(to_type)
    snapshot.assert_match(expr.compile(), "out.sql")


@pytest.mark.parametrize(
    "column",
    [
        "id",
        "bool_col",
        "tinyint_col",
        "smallint_col",
        "int_col",
        "bigint_col",
        "float_col",
        "double_col",
        "date_string_col",
        "string_col",
        "timestamp_col",
        "year",
        "month",
    ],
)
def test_noop_cast(alltypes, column, snapshot):
    col = alltypes[column]
    result = col.cast(col.type())
    assert result.equals(col)
    snapshot.assert_match(result.compile(), "out.sql")


def test_timestamp_cast(alltypes, snapshot):
    target = dt.Timestamp(nullable=False)
    result1 = alltypes.timestamp_col.cast(target)
    result2 = alltypes.int_col.cast(target)

    assert isinstance(result1, ir.TimestampColumn)
    assert isinstance(result2, ir.TimestampColumn)

    snapshot.assert_match(result1.compile(), "out1.sql")
    snapshot.assert_match(result2.compile(), "out2.sql")


def test_timestamp_now(con, snapshot):
    expr = ibis.now()
    snapshot.assert_match(con.compile(expr), "out.sql")


@pytest.mark.parametrize("unit", ["y", "m", "d", "w", "h", "minute"])
def test_timestamp_truncate(con, unit, snapshot):
    stamp = ibis.timestamp("2009-05-17 12:34:56")
    expr = stamp.truncate(unit)
    snapshot.assert_match(con.compile(expr), "out.sql")


@pytest.mark.parametrize(("value", "expected"), [(0, None), (5.5, 5.5)])
def test_nullifzero(con, value, expected):
    with pytest.warns(FutureWarning):
        result = con.execute(L(value).nullifzero())
    if expected is None:
        assert pd.isnull(result)
    else:
        assert result == expected


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        (L(None).isnull(), True),
        (L(1).isnull(), False),
        (L(None).notnull(), False),
        (L(1).notnull(), True),
    ],
)
def test_isnull_notnull(con, expr, expected):
    assert con.execute(expr) == expected


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        (ibis.coalesce(5, None, 4), 5),
        (ibis.coalesce(ibis.NA, 4, ibis.NA), 4),
        (ibis.coalesce(ibis.NA, ibis.NA, 3.14), 3.14),
    ],
)
def test_coalesce(con, expr, expected):
    assert con.execute(expr) == expected


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        (ibis.NA.fillna(5), 5),
        (L(5).fillna(10), 5),
        (L(5).nullif(5), None),
        (L(10).nullif(5), 10),
    ],
)
def test_fillna_nullif(con, expr, expected):
    result = con.execute(expr)
    if expected is None:
        assert pd.isnull(result)
    else:
        assert result == expected


@pytest.mark.parametrize(
    ("value", "expected"),
    [
        (L("foo_bar"), "String"),
        (L(5), "UInt8"),
        (L(1.2345), "Float64"),
        (L(datetime(2015, 9, 1, hour=14, minute=48, second=5)), "DateTime"),
        (L(date(2015, 9, 1)), "Date"),
        param(
            ibis.NA,
            "Null",
            marks=pytest.mark.xfail(
                raises=AssertionError,
                reason=(
                    "Client/server version mismatch not handled in the "
                    "clickhouse driver"
                ),
            ),
        ),
    ],
)
def test_typeof(con, value, expected):
    assert con.execute(value.typeof()) == expected


@pytest.mark.parametrize(("value", "expected"), [("foo_bar", 7), ("", 0)])
def test_tuple_string_length(con, value, expected):
    assert con.execute(L(value).length()) == expected


@pytest.mark.parametrize(
    ("op", "expected"),
    [
        (methodcaller("substr", 0, 3), "foo"),
        (methodcaller("substr", 4, 3), "bar"),
        (methodcaller("substr", 1), "oo_bar"),
    ],
)
def test_string_substring(con, op, expected):
    value = L("foo_bar")
    assert con.execute(op(value)) == expected


def test_string_column_substring(con, alltypes, snapshot):
    expr = alltypes.string_col.substr(2)
    snapshot.assert_match(expr.compile(), "out1.sql")
    assert len(con.execute(expr))

    expr = alltypes.string_col.substr(0, 3)
    snapshot.assert_match(expr.compile(), "out2.sql")
    assert len(con.execute(expr))


def test_string_reverse(con):
    assert con.execute(L("foo").reverse()) == "oof"


def test_string_upper(con):
    assert con.execute(L("foo").upper()) == "FOO"


def test_string_lower(con):
    assert con.execute(L("FOO").lower()) == "foo"


def test_string_length(con):
    assert con.execute(L("FOO").length()) == 3


@pytest.mark.parametrize(
    ("value", "op", "expected"),
    [
        (L("foobar"), methodcaller("contains", "bar"), True),
        (L("foobar"), methodcaller("contains", "foo"), True),
        (L("foobar"), methodcaller("contains", "baz"), False),
        (L("100%"), methodcaller("contains", "%"), True),
        (L("a_b_c"), methodcaller("contains", "_"), True),
    ],
)
def test_string_contains(con, op, value, expected):
    assert con.execute(op(value)) == expected


def test_re_replace(con):
    expr1 = L("Hello, World!").re_replace(".", r"\0\0")
    expr2 = L("Hello, World!").re_replace("^", "here: ")

    assert con.execute(expr1) == "HHeelllloo,,  WWoorrlldd!!"
    assert con.execute(expr2) == "here: Hello, World!"


@pytest.mark.parametrize(
    ("value", "expected"),
    [(L("a"), 0), (L("b"), 1), (L("d"), -1)],  # TODO: what's the expected?
)
def test_find_in_set(con, value, expected):
    vals = list("abc")
    expr = value.find_in_set(vals)
    assert con.execute(expr) == expected


def test_string_column_find_in_set(con, alltypes, snapshot):
    s = alltypes.string_col
    vals = list("abc")

    expr = s.find_in_set(vals)
    snapshot.assert_match(expr.compile(), "out.sql")
    assert len(con.execute(expr))


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        (L("foobar").find("bar"), 3),
        (L("foobar").find("baz"), -1),
        (L("foobar").like("%bar"), True),
        (L("foobar").like("foo%"), True),
        (L("foobar").like("%baz%"), False),
        (L("foobar").like(["%bar"]), True),
        (L("foobar").like(["foo%"]), True),
        (L("foobar").like(["%baz%"]), False),
        (L("foobar").like(["%bar", "foo%"]), True),
        (L("foobarfoo").replace("foo", "H"), "HbarH"),
    ],
)
def test_string_find_like(con, expr, expected):
    assert con.execute(expr) == expected


def test_string_column_like(con, alltypes, snapshot):
    expr = alltypes.string_col.like("foo%")
    snapshot.assert_match(expr.compile(), "out1.sql")
    assert len(con.execute(expr))

    expr = alltypes.string_col.like(["foo%", "%bar"])
    snapshot.assert_match(expr.compile(), "out2.sql")
    assert len(con.execute(expr))


def test_string_column_find(con, alltypes, snapshot):
    s = alltypes.string_col

    expr = s.find("a")
    snapshot.assert_match(expr.compile(), "out1.sql")
    assert len(con.execute(expr))

    expr = s.find(s)
    snapshot.assert_match(expr.compile(), "out2.sql")
    assert len(con.execute(expr))


@pytest.mark.parametrize(
    "call",
    [
        param(methodcaller("log"), id="log"),
        param(methodcaller("log2"), id="log2"),
        param(methodcaller("log10"), id="log10"),
        param(methodcaller("round"), id="round"),
        param(methodcaller("round", 0), id="round_0"),
        param(methodcaller("round", 2), id="round_2"),
        param(methodcaller("exp"), id="exp"),
        param(methodcaller("abs"), id="abs"),
        param(methodcaller("ceil"), id="ceil"),
        param(methodcaller("sqrt"), id="sqrt"),
        param(methodcaller("sign"), id="sign"),
    ],
)
def test_translate_math_functions(con, alltypes, call, snapshot):
    expr = call(alltypes.double_col)
    snapshot.assert_match(expr.compile(), "out.sql")
    assert len(con.execute(expr))


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        pytest.param(L(-5).abs(), 5, id="abs_neg"),
        pytest.param(L(5).abs(), 5, id="abs"),
        pytest.param(L(5.5).round(), 6.0, id="round"),
        pytest.param(L(5.556).round(2), 5.56, id="round_places"),
        pytest.param(L(5.556).ceil(), 6.0, id="ceil"),
        pytest.param(L(5.556).floor(), 5.0, id="floor"),
        pytest.param(L(5.556).sign(), 1, id="sign"),
        pytest.param(L(-5.556).sign(), -1, id="sign_neg"),
        pytest.param(L(0).sign(), 0, id="sign_zero"),
        pytest.param(L(5.556).sqrt(), math.sqrt(5.556), id="sqrt"),
        pytest.param(L(5.556).log(2), math.log(5.556, 2), id="log2_arg"),
        pytest.param(L(5.556).log2(), math.log(5.556, 2), id="log2"),
        pytest.param(L(5.556).log10(), math.log10(5.556), id="log10"),
        # clickhouse has different functions for exp/ln that are faster
        # than the defaults, but less precise
        #
        # we can't use the e() function as it still gives different results
        # from `math.exp`
        pytest.param(
            L(5.556).exp().round(8),
            round(math.exp(5.556), 8),
            id="exp",
        ),
        pytest.param(
            L(5.556).ln().round(7),
            round(math.log(5.556), 7),
            id="ln",
        ),
    ],
)
def test_math_functions(con, expr, expected):
    assert con.execute(expr) == expected


def test_greatest_least(con, alltypes, snapshot):
    expr = ibis.greatest(alltypes.int_col, 10)
    snapshot.assert_match(expr.compile(), "out1.sql")
    assert len(con.execute(expr))

    expr = ibis.greatest(alltypes.int_col, alltypes.bigint_col)
    snapshot.assert_match(expr.compile(), "out2.sql")
    assert len(con.execute(expr))

    expr = ibis.least(alltypes.int_col, 10)
    snapshot.assert_match(expr.compile(), "out3.sql")
    assert len(con.execute(expr))

    expr = ibis.least(alltypes.int_col, alltypes.bigint_col)
    snapshot.assert_match(expr.compile(), "out4.sql")
    assert len(con.execute(expr))


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        (L("abcd").re_search("[a-z]"), True),
        (L("abcd").re_search(r"[\d]+"), False),
        (L("1222").re_search(r"[\d]+"), True),
    ],
)
def test_regexp(con, expr, expected):
    assert con.execute(expr) == expected


@pytest.mark.parametrize(
    ("expr", "expected"),
    [
        param(L("abcd").re_extract("([a-z]+)", 0), "abcd", id="simple"),
        # (L('abcd').re_extract('(ab)(cd)', 1), 'cd'),
        # valid group number but no match => None
        param(L("abcd").re_extract(r"(\\d)", 0), None, id="valid_group_no_match"),
        # match but not a valid group number => NULL
        param(L("abcd").re_extract("abcd", 3), None, id="invalid_group_match"),
    ],
)
def test_regexp_extract(con, expr, expected):
    assert con.execute(expr) == expected


def test_column_regexp_extract(con, alltypes, snapshot):
    expr = alltypes.string_col.re_extract(r"[\d]+", 3)
    snapshot.assert_match(expr.compile(), "out.sql")
    assert len(con.execute(expr))


def test_column_regexp_replace(con, alltypes, snapshot):
    expr = alltypes.string_col.re_replace(r"[\d]+", "aaa")
    snapshot.assert_match(expr.compile(), "out.sql")
    assert len(con.execute(expr))


def test_numeric_builtins_work(alltypes, df):
    expr = alltypes.double_col
    result = expr.execute()
    expected = df.double_col.fillna(0)
    tm.assert_series_equal(result, expected)


def test_null_column(alltypes):
    t = alltypes
    nrows = t.count().execute()
    expr = t.mutate(na_column=ibis.NA).na_column
    result = expr.execute()
    expected = pd.Series([None] * nrows, name="na_column")
    tm.assert_series_equal(result, expected)


def test_literal_none_to_nullable_column(alltypes):
    # GH: 2985
    t = alltypes
    nrows = t.count().execute()
    expr = t.mutate(
        ibis.literal(None, dt.String(nullable=True)).name("nullable_string_column")
    )
    result = expr["nullable_string_column"].execute()
    expected = pd.Series([None] * nrows, name="nullable_string_column")
    tm.assert_series_equal(result, expected)


def test_timestamp_from_integer(con, alltypes, snapshot):
    # timestamp_col has datetime type
    expr = alltypes.int_col.to_timestamp()
    snapshot.assert_match(expr.compile(), "out.sql")
    assert len(con.execute(expr))


def test_count_distinct_with_filter(alltypes):
    expr = alltypes.string_col.nunique(where=alltypes.string_col.cast("int64") > 1)
    result = expr.execute()
    expected = alltypes.string_col.execute()
    expected = expected[expected.astype("int64") > 1].nunique()
    assert result == expected


@pytest.mark.parametrize(
    ("sep", "where_case"),
    [
        param(",", None, id="comma_none"),
        param("-", None, id="minus_none"),
        param(",", 0, id="comma_zero"),
    ],
)
def test_group_concat(alltypes, sep, where_case, snapshot):
    where = None if where_case is None else alltypes.bool_col == where_case
    expr = alltypes.string_col.group_concat(sep, where)
    snapshot.assert_match(expr.compile(), "out.sql")


def test_hash(alltypes, snapshot):
    expr = alltypes.string_col.hash()
    snapshot.assert_match(expr.compile(), "out.sql")


def test_udf_in_array_map(alltypes):
    @udf.scalar.builtin(name="plus")
    def my_add(a: int, b: int) -> int:
        ...

    n = 5
    expr = (
        alltypes[alltypes.int_col == 1]
        .limit(n)
        .int_col.collect()
        .map(lambda x: my_add(x, 1))
    )
    result = expr.execute()
    assert result == [2] * n


def test_udf_in_array_filter(alltypes):
    @udf.scalar.builtin(name="equals")
    def my_eq(a: int, b: int) -> bool:
        ...

    expr = alltypes.int_col.collect().filter(lambda x: my_eq(x, 1))
    result = expr.execute()
    assert set(result) == {1}
