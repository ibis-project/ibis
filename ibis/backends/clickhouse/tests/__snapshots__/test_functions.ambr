# name: test_cast_double_col[float]
  '`double_col`'
# ---
# name: test_cast_double_col[int16]
  'CAST(`double_col` AS Nullable(Int16))'
# ---
# name: test_cast_string_col[!string]
  'CAST(`string_col` AS String)'
# ---
# name: test_column_regexp_extract
  "if(match(CAST(`string_col` AS String), '[\\d]+'), if(CAST(nullIf(3, 0) AS Nullable(Int64)) IS NULL, CAST(`string_col` AS String), CAST(extractAll(CAST(`string_col` AS String), '[\\d]+') AS Array(Nullable(String)))[CAST(nullIf(3, 0) AS Nullable(Int64))]), NULL)"
# ---
# name: test_greatest_least
  'greatest(`int_col`, 10)'
# ---
# name: test_group_concat[minus_none]
  "CASE WHEN empty(groupArray(`string_col`)) THEN NULL ELSE arrayStringConcat(groupArray(`string_col`), '-') END"
# ---
# name: test_noop_cast[double_col]
  '`double_col`'
# ---
# name: test_noop_cast[smallint_col]
  '`smallint_col`'
# ---
# name: test_noop_cast[year]
  '`year`'
# ---
# name: test_timestamp_truncate[y]
  "SELECT toStartOfYear(toDateTime('2009-05-17T12:34:56')) AS `TimestampTruncate(Timestamp('2009-05-17 12:34:56'))`"
# ---
# name: test_translate_math_functions[log2]
  'log2(`double_col`)'
# ---
# name: test_translate_math_functions[log]
  'log(`double_col`)'
# ---
# name: test_translate_math_functions[round]
  'round(`double_col`)'
# ---
# name: test_translate_math_functions[round_2]
  'round(`double_col`, 2)'
# ---
